### ISSUE TITLE
Improve`compose_result`

### ISSUE THREAD

**Issue by sobolevn** [dry-python/returns#502]
We need to add:
- [ ] HKT method `compose_result()`
- [ ] pointfree helper
- [ ] type tests for all existing types with `.compose_result()`
- [ ] type tests for pointfree and method

**Comment by thepabloaguilar** [comment]
Is not just one method `compose_result`, in fact it's:

* `compose_result` -> `IOResult`
* `compose_future_result` -> `FutureResult`
* `compose_requires_context_io_result` -> `RequiresContextIOResult`
* `compose_requires_context_future_result` -> `RequiresContextFutureResult`

**Comment by thepabloaguilar** [comment]
Why `compose_result` has this signature?

```python
class IOResultLikeN:
    @abstractmethod
    def compose_result(
        self: _IOResultLikeType,
        function: Callable[
            ['Result[_FirstType, _SecondType]'],
            KindN[_IOResultLikeType, _UpdatedType, _SecondType, _ThirdType],
        ],
    ) -> KindN[_IOResultLikeType, _UpdatedType, _SecondType, _ThirdType]:
        """Allows to compose the unrelying ``Result`` with a function."""
```

Instead of:

```python
class IOResultLikeN:
    @abstractmethod
    def compose_result(
        self: _IOResultLikeType,
        function: Callable[
            [KindN[_ResultLikeKind, _FirstType, _SecondType, _ThirdType]], # HERE
            KindN[_IOResultLikeType, _UpdatedType, _SecondType, _ThirdType],
        ],
    ) -> KindN[_IOResultLikeType, _UpdatedType, _SecondType, _ThirdType]:
        """Allows to compose the unrelying ``Result`` with a function."""
```

---

Another question is why just the first value (`_FirstType` -> `_UpdatedType`) changes?? Because using `compose_result` I can return any type of `IOResult`!!

**Comment by sobolevn** [comment]
The idea behind this method is that sometimes we need to work with `pure` `Result` value.
Let's say we have a `FutureResult` object. We cannot not compose it with a function that takes just `Result`.
There's literally no other way. 

That's how `compose_result` helps. 

It was originally created for https://github.com/dry-python/returns/blob/master/returns/_generated/pipeline/managed.py#L114

**Comment by sobolevn** [comment]
We can probably add:
- `compose_ioresult` for `IOResult+` types
- `compose_future_result` for `FutureResult+` types
- `compose_maybe` for `Maybe+` types

**Comment by thepabloaguilar** [comment]
> That's how `compose_result` helps.

Yes, but my question is about the "`hardcode`" type. I think the `compose_result` method should use `KindN` in the callable signature!
Maybe my confusion is because `IOResult` is a container of container, it stores a `Result` inside of it.
The users will not be able to use their own `Result` containers, we're limiting to use _returns_ `Result` in our `IOResultLikeN` interface.

**Comment by sobolevn** [comment]
No, that's exactly the point of `compose_result` to work with `Result`! 🙂 

When some type is based on `Result`, we would sometimes need to use it. With`compose_result`!

**Comment by thepabloaguilar** [comment]
> We can probably add:
> 
> * `compose_ioresult` for `IOResult+` types
> * `compose_future_result` for `FutureResult+` types
> * `compose_maybe` for `Maybe+` types

`compose_result` naturally will be for `IOResult+`, we just have the `compose_result(...)` method in the `IOResultLikeN` interface!
